gcloud-java-compute/src/main/java/com/google/gcloud/compute/AddressInfo.java

```diff
@@ -0,0 +1,576 @@
+      return this;
+    }
+
+    @Override
+    public AddressInfo build() {
+      return new AddressInfo(this);
+    }
+  }
+
+  AddressInfo(BuilderImpl builder) {
+    address = builder.address;
+    creationTimestamp = builder.creationTimestamp;
+    description = builder.description;
+    id = builder.id;
+    addressId = checkNotNull(builder.addressId);
+    status = builder.status;
+    usage = builder.usage;
+  }
+
+  /**
+   * Returns the static external IP address represented by this object.
+   */
+  public String address() {
+    return address;
+  }
+
+  /**
+   * Returns the creation timestamp in milliseconds since epoch.
+   */
+  public Long creationTimestamp() {
+    return creationTimestamp;
+  }
+
+  /**
+   * Returns an optional textual description of the address.
+   */
+  public String description() {
+    return description;
+  }
+
+  /**
+   * Returns the unique identifier for the address; defined by the service.
+   */
+  public String id() {
+    return id;
+  }
+
+  /**
+   * Returns the address identity. Returns {@link GlobalAddressId} for a global address, returns
+   * {@link RegionAddressId} for a region address.
+   */
+  @SuppressWarnings("unchecked")
+  public <T extends AddressId> T addressId() {
+    return (T) addressId;
+  }
+
+  /**
+   * Returns the status of the address.
+   */
+  public Status status() {
+    return status;
+  }
+
+  /**
+   * Returns the usage information of the address. Returns a {@link InstanceUsage} object for region
+   * addresses that are assigned to VM instances. Returns a {@link RegionForwardingUsage} object for
+   * region addresses assigned to region forwarding rules. Returns a {@link GlobalForwardingUsage}
+   * object for global addresses assigned to global forwarding rules. Returns {@code null} if the
+   * address is not in use.
+   */
+  @SuppressWarnings("unchecked")
+  public <T extends Usage> T usage() {
+    return (T) usage;
+  }
+
+  /**
+   * Returns a builder for the {@code AddressInfo} object.
+   */
+  public Builder toBuilder() {
+    return new BuilderImpl(this);
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("address", address)
+        .add("creationTimestamp", creationTimestamp)
+        .add("description", description)
+        .add("id", id)
+        .add("addressId", addressId)
+        .add("status", status)
+        .add("usage", usage)
+        .toString();
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(address, creationTimestamp, description, id, addressId, status, usage);
+  }
+
+  @Override
+  public boolean equals(Object obj) {
+    return obj != null
+        && obj.getClass().equals(AddressInfo.class)
+        && Objects.equals(toPb(), ((AddressInfo) obj).toPb());
+  }
+
+  AddressInfo setProjectId(String projectId) {
+    Builder builder = toBuilder();
+    AddressId addressIdWithProject;
+    if (addressId instanceof RegionAddressId) {
```

Comment:
Similar issue as before. I think it is a problem when we are accepting an interface but then are depending on specific implementation.

You should try to avoid it however if too difficult then you should avoid accepting something that you are not  expecting  (e.g. check on `Builder.addressId` or change `Builder.addressId` to accept the 2 specific implementations).